### DISCUSSION TITLE
miRge-buildについて

### DISCUSSION THREAD

**Discussion by TomokiMotegi** [nig-sc/nigsc_homepage2#522] [Q&A]
いつも素晴らしい環境の提供ありがとうございます。

表題にもありますようにmiRge-buildを用いてmiRge3用のライブラリを作成しようと試みています。
しかしこれを/lustre7/singularity/images/biotools/m/mirge-build\:0.0.1--pyh3252c3a_0のイメージから動かそうとすると、
`AttributeError: 'NoneType' object has no attribute 'span'
`
と出て止まってしまいます。
本ソフトのgithubに行くとpython3.8への対応をしているとなっており、このイメージ上ではpython3.7が動いているためのエラーかと思いますが、他のイメージを作成することはできませんでしょうか？

以下が私が実行したコードとなります。
`singularity exec /lustre7/singularity/images/biotools/m/mirge-build\:0.0.1--pyh3252c3a_0 miRge-build -g ~/share/CanFam3.1_transcript/Canis_familiaris.CanFam3.1.dna.toplevel.fa -mmf ~/project/Vicky_Tufts/miRge3/cfa3_lib/miRBase_cfa_mature_miRNA.fa -hmf ~/project/Vicky_Tufts/miRge3/cfa3_lib/miRBase_cfa_hairpin_miRNA.fa -mtf ~/project/Vicky_Tufts/miRge3/cfa3_lib/canFam3-tRNAs.fa -ptf ~/project/Vicky_Tufts/miRge3/cfa3_lib/pre_tRNA.fa -snorf ~/project/Vicky_Tufts/miRge3/cfa3_lib/rnacentral_cfa_snoRNA.nowrap.fa -rrf ~/project/Vicky_Tufts/miRge3/cfa3_lib/rnacentral_cfa_rRNA.nowrap.fa -ncof ~/project/Vicky_Tufts/miRge3/cfa3_lib/rnacentral_cfa_other_non_cording_RNA.nowrap.fa -mrf ~/share/CanFam3.1_transcript/Canis_familiaris.CanFam3.1.cdna.all.fa -agff ~/project/Vicky_Tufts/miRge3/cfa3_lib/cfa.gff3 -db miRBase -on cfa3 -cpu 10 `

お忙しいところ恐縮ですが、ご回答いただければ幸いです。

茂木

**Comment by oogasawa** [comment]
以下に作成してみました。https://github.com/nig-sc/apptainer_mirge3
特定のバージョンのPython (3.8), numpy (1.19.x)が必要なようなのでapptainer (旧singularity)コンテナを使ってバージョンを正確にコントロールした環境を作ったほうが良いようでした。

動作確認をちゃんとしていないので問題があるようでしたらお知らせください。（うまく動くようでしたらREADME.mdにまともな例を書き足しておいていただけますと助かります。）

**Reply by TomokiMotegi** [reply]
お忙しい中のご対応ありがとうございます。
作成してみたところmirge3.sifは動きました。
miRage-buildは作成できたのですが、
- UndefinedVar: Usage of undefined variable '$HOME' (line 6)
となっており、$HOMEの定義が無い点が少し気になりました。
また名前の定義が違っており、正しくは以下の通りでした
旧 ’apptainer exec mirge-build.sif mirge-build --version'
正 ’apptainer exec mirge-build.sif miRge-build --version'
ただ、これでバージョン情報は表示されます。

上記は些末事なのですが、実際にライブラリー作成を行ってみると以下のエラーが出ます。
実行コード
`apptainer exec mirge-build.sif miRge-build -g ./cfa3_lib/CanFam3.1.fa -mmf ./cfa3_lib/miRBase_cfa_mature_miRNA.fa -hmf ./cfa3_lib/miRBase_cfa_hairpin_miRNA.fa -mtf ./cfa3_lib/canFam3-tRNAs.fa -ptf ./cfa3_lib/pre_tRNA.fa -snorf ./cfa3_lib/rnacentral_cfa_snoRNA.nowrap.fa -rrf ./cfa3_lib/rnacentral_cfa_rRNA.nowrap.fa -ncof ./cfa3_lib/rnacentral_cfa_other_non_cording_RNA.nowrap.fa -mrf ./cfa3_lib/mrna.fa -agff ./cfa3_lib/cfa.gff3 -db miRBase -on cfa3_test -cpu 10`
エラー
```
Traceback (most recent call last):
  File "/opt/venv/bin/miRge-build", line 8, in <module>
    sys.exit(main())
  File "/opt/venv/lib/python3.8/site-packages/mirge_build/__main__.py", line 27, in main
    outlog.write("bowtie error!: bowtie, command not found \nUse argument -pbwt <name of the directory>\n")
NameError: name 'outlog' is not defined
```
outlogが入っていないのかと思って、Dockerfileのpipの項目に追加してみたのですが、定義自体がされておらず上手く動きません。
何か良い解決策はありますでしょうか？

茂木
